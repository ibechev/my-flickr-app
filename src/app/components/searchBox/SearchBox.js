import React from "react";
import PropTypes from "prop-types";

import Tag from "../tag/Tag";
import Button from "../button/Button";

export const SearchBox = ({
  inputFocus,
  inputValue,
  error,
  tags,
  refs,
  ...handle
}) => {
  const focusClass = inputFocus ? "focus" : "";
  const errorClass = error ? "error" : "";
  const placeholder = "... Enter tags here. Separate with space or comma";

  return (
    <section
      id="search-box"
      className={`search-box ${errorClass} ${focusClass}`}
      onClick={handle.searchBoxFocus}
      ref={refs.searchBox}
      data-test="search-box"
    >
      {error && (
        <span className="error" data-test="input-error">
          {error}
        </span>
      )}

      <form action="" onSubmit={e => handle.addTag(e, inputValue.trim())}>
        <section className="tags">
          <ul className="tags-wrapper">
            {tags.map((tag, i) => (
              <Tag
                key={tag}
                index={i}
                value={tag}
                removeTag={handle.removeTag}
              />
            ))}

            <li className="tags-input-wrapper">
              <input
                id="tags-input"
                type="text"
                className="tags-input"
                size="4"
                autoComplete="off"
                onKeyDown={handle.keyDown}
                onChange={handle.change}
                autoFocus
                ref={refs.inputEl}
                value={inputValue}
              />
            </li>
          </ul>

          {!tags.length && !inputValue.length && (
            <span className="placeholder" data-test="tags-placeholder">
              {placeholder}
            </span>
          )}
        </section>

        <section className="controls">
          <Button
            className="clear"
            onClick={handle.removeAllTags}
            title="Clear all tags"
            type="button"
            data-test="btn-clear-tags"
          >
            <i className="far fa-trash-alt" />
            Clear
          </Button>
        </section>
      </form>
    </section>
  );
};

SearchBox.propTypes = {
  tags: PropTypes.arrayOf(PropTypes.string),
  inputValue: PropTypes.string,
  inputFocus: PropTypes.bool.isRequired,
  error: PropTypes.string,
  refs: PropTypes.shape({
    searchBox: PropTypes.shape({
      current: PropTypes.node.isRequired.isRequired
    }).isRequired,
    inputEl: PropTypes.shape({
      current: PropTypes.node.isRequired.isRequired
    }).isRequired
  }),
  handle: PropTypes.shape({
    searchBoxFocus: PropTypes.func.isrequired,
    keyDown: PropTypes.func.isrequired,
    addTag: PropTypes.func.isrequired,
    change: PropTypes.func.isrequired,
    search: PropTypes.func.isrequired,
    removeTag: PropTypes.func.isrequired,
    removeAllTags: PropTypes.func.isrequired
  })
};

export default SearchBox;
