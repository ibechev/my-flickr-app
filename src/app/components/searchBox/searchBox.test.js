import React, { createRef } from "react";
import renderer from "react-test-renderer";
import { SearchBox } from "./SearchBox";
import { propTypeErrors } from "../../../../tests/testUtils";
import { testTags } from "../../../../tests/testValues";
import { mount } from "enzyme";

describe("Expect <SearchBox />", () => {
  const defaultProps = {
    inputValue: "",
    inputFocus: true,
    error: "",
    tags: [],
    refs: { searchBox: createRef(), inputEl: createRef() },
    searchBoxFocus: () => {},
    keyDown: () => {},
    addTag: () => {},
    change: () => {},
    search: () => {},
    removeTag: () => {},
    removeAllTags: () => {}
  };

  const setup = (props = {}) => {
    const wrapper = mount(<SearchBox {...defaultProps} {...props} />);
    return wrapper;
  };

  it("to render correctly", () => {
    const wrapper = renderer.create(<SearchBox {...defaultProps} />).toJSON();
    expect(wrapper).toMatchSnapshot();
  });

  it("to not throw error with correct props", () => {
    propTypeErrors(<SearchBox {...defaultProps} />, defaultProps);
  });

  it("to not display any <Tag /> conponent, if 'state.tags' is empty", () => {
    const wrapper = setup(defaultProps);
    expect(wrapper.find("Tag")).toHaveLength(0);
  });

  it("to display number of <Tag /> components, equal of the number of tags in 'props.tags'", () => {
    const wrapper = setup({ tags: [...testTags] });
    expect(wrapper.find("Tag")).toHaveLength(testTags.length);
  });

  it("to display <span data-test='tags-placeholder' /> only, if 'props.tags' is empty and/or 'props.inputValue.length' eauals 0", () => {
    const testPropsOne = {
      tags: [...testTags]
    };
    const testPropsTwo = {
      inputValue: "abc"
    };
    const testPropsThree = {
      tags: [...testTags],
      inputValue: "abc"
    };
    const wrapper = setup();
    const wrapperOne = setup(testPropsOne);
    const wrapperTwo = setup(testPropsTwo);
    const wrapperThree = setup(testPropsThree);

    const spanName = "span[data-test='tags-placeholder']";

    expect(wrapper.find(spanName)).toHaveLength(1);
    expect(wrapperOne.find(spanName)).toHaveLength(0);
    expect(wrapperTwo.find(spanName)).toHaveLength(0);
    expect(wrapperThree.find(spanName)).toHaveLength(0);
  });

  it("to add class 'focus' to <section data-test='search-box' />, if 'props.inFocus' equals 'true'", () => {
    const wrapperWithFocus = setup();
    const wrapperWithoutFocus = setup({ inputFocus: false });

    const sectionName = "section[data-test='search-box']";
    const focusClass = "focus";

    expect(wrapperWithFocus.find(sectionName).hasClass(focusClass)).toEqual(
      true
    );
    expect(wrapperWithoutFocus.find(sectionName).hasClass(focusClass)).toEqual(
      false
    );
  });

  it("to display <span data-test='input-error' /> and the error text inside, only if 'props.error' is not an empty string", () => {
    const testProps = {
      error: "Some error message"
    };

    const wrapper = setup();
    const spanName = "span[data-test='input-error']";
    expect(wrapper.find(spanName)).toHaveLength(0);

    const wrapperWithError = setup(testProps);
    expect(wrapperWithError.find(spanName)).toHaveLength(1);
    expect(wrapperWithError.find(spanName).text()).toEqual(testProps.error);
  });

  it("to call 'props.removeAllTags', when <button dats-test='clear-tags' is clicked", () => {
    const removeAllTags = jest.fn();
    const wrapper = setup({ removeAllTags });

    wrapper.find("button[data-test='btn-clear-tags']").simulate("click");
    expect(removeAllTags).toHaveBeenCalledTimes(1);
  });

  it("to call 'props.addTag' only when 'props.inputValue' is not an empty string AND 'props.error' IS an empty string", () => {});

  describe("to call 'props.removeTag' only when", () => {
    it();
  });

  // it("to remove all tags when button 'Clear' is clicked", () => {
  //   const testState = {
  //     tags: [...testTags]
  //   };
  //   const wrapper = setup(defaultProps, testState);
  //   wrapper.find("button.clear").simulate("click");
  //   expect(wrapper.state("tags")).toHaveLength(0);
  // });

  // it("to add anothet tag when comma or space is pressed and the 'input' is not empty", () => {
  //   const testState = {
  //     tags: [...testTags]
  //   };
  //   const valueOne = "sandwich";
  //   const valueTwo = "alpha";
  //   const wrapper = setup(defaultProps, testState);
  //   const input = wrapper.find("input.tags-input");

  //   input.simulate("focus");
  //   input.simulate("keyDown", {
  //     keyCode: 32, // space
  //     target: { value: valueOne },
  //     preventDefault() {}
  //   });

  //   const tagsResultOne = [...wrapper.state("tags")];
  //   expect(tagsResultOne).toHaveLength(testState.tags.length + 1);
  //   expect(tagsResultOne.pop()).toEqual(valueOne);

  //   input.simulate("keyDown", {
  //     keyCode: 188, // comma
  //     target: { value: valueTwo },
  //     preventDefault() {}
  //   });

  //   const tagsResultTwo = [...wrapper.state("tags")];
  //   expect(tagsResultTwo).toHaveLength(testState.tags.length + 2);
  //   expect(tagsResultTwo.pop()).toEqual(valueTwo);
  // });

  // it("to set focus on the 'input' and set 'state.inFocus' to 'true' when user clicks on <section class='search-box'", () => {
  //   const wrapper = setup(defaultProps);
  //   wrapper.find("section.search-box").simulate("click", { target: {} });
  //   expect(wrapper.find("section.search-box").hasClass("focus")).toEqual(true);
  //   expect(wrapper.state("inFocus")).toEqual(true);
  // });

  // it("to remove single tag from 'state.tags', each time backspace is pressed", () => {
  //   const testState = {
  //     tags: [...testTags]
  //   };
  //   const wrapper = setup(defaultProps, testState);
  //   const input = wrapper.find("input.tags-input");

  //   input.simulate("focus");
  //   input.simulate("keyDown", {
  //     keyCode: 8, // backspace
  //     target: { value: "" }
  //   });
  //   const tagsResult = [...wrapper.state("tags")];
  //   expect(tagsResult).toHaveLength(testState.tags.length - 1);
  //   expect(tagsResult[0]).toEqual(testState.tags[0]);
  //   expect(tagsResult[1]).toEqual(testState.tags[1]);
  // });
});
