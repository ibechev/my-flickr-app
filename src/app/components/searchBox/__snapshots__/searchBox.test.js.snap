// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expect <SearchBox /> to render correctly 1`] = `
<section
  className="search-box  focus"
  id="search-box"
  onClick={[Function]}
>
  
  <form
    action=""
    onSubmit={[Function]}
  >
    <section
      className="tags"
    >
      <ul
        className="tags-wrapper"
      >
        <li
          className="tags-input-wrapper"
        >
          <input
            autoComplete="off"
            autoFocus={true}
            className="tags-input"
            id="tags-input"
            onChange={[Function]}
            onKeyDown={[Function]}
            size="4"
            type="text"
            value=""
          />
        </li>
      </ul>
      <span
        className="placeholder"
      >
        ... Enter tags here. Separate with space or comma
      </span>
    </section>
    <section
      className="controls"
    >
      <button
        className="btn clear"
        onClick={[Function]}
        title="Clear all tags"
        type="button"
      >
        <i
          className="far fa-trash-alt"
        />
        Clear
      </button>
    </section>
  </form>
</section>
`;
